### PR TITLE
add client credential flow

### LINKED ISSUE #120: add local SMTP

### LINKED ISSUE #121: Remove EntityFramework library
https://github.com/SergiLomtadze/SERGI_LOMTADZE_INDIVIDUAL_MENTORSHIP/blob/8ea437199dfe40cd2e1660d76ba9a74f48bb034d/src/ExadelMentorship.DataAccess/ExadelMentorship.DataAccess.csproj#L8

### PR DESCRIPTION
**Pull request by SergiLomtadze**
close #121 
close #120

### CODE REVIEW

File: src/ExadelMentorship.Auth/IdentityConfiguration.cs [hunk 1]
```diff
@@ -21,7 +19,8 @@ public class IdentityConfiguration
         public static IEnumerable<ApiScope> Scopes => new[]
         {
             new ApiScope("WebApi.read"),
-            new ApiScope("role")
+            new ApiScope("role"),
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Auth/IdentityConfiguration.cs:22]: Probably not needed. Usually, we use "role"s as claims, not scopes.

File: src/ExadelMentorship.Auth/IdentityConfiguration.cs [hunk 2]
```diff
@@ -35,6 +34,17 @@ public class IdentityConfiguration
...
                 AllowedCorsOrigins = {"https://localhost:7066"},
+            },
+
+            new Client
+            {
+              ClientId = "reportJob",
+              AllowedGrantTypes = GrantTypes.ClientCredentials,
+              ClientSecrets =
+              {
+                new Secret("secret".Sha256())
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Auth/IdentityConfiguration.cs:45]: Keep a secret secret

File: src/ExadelMentorship.Jobs/ReportJob.cs [hunk 3]
```diff
@@ -42,14 +45,29 @@ protected async override Task ExecuteAsync(CancellationToken stoppingToken)
         //System.NotSupportedException: 'Only public methods can be invoked in the background'
         public async Task Execute(ReportUser reportUser)
         {
+            var client = new HttpClient
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Jobs/ReportJob.cs:48]: Use IHttpClientFactory way. Keep the url in the config.

File: src/ExadelMentorship.Jobs/ReportJob.cs [hunk 4]
```diff
@@ -42,14 +45,29 @@ protected async override Task ExecuteAsync(CancellationToken stoppingToken)
...
+            {
+            };
+
+            var disco = await client.GetDiscoveryDocumentAsync("https://localhost:7046");
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Jobs/ReportJob.cs:52]: Nice naming choice

File: src/ExadelMentorship.Jobs/ReportJob.cs [hunk 5]
```diff
@@ -42,14 +45,29 @@ protected async override Task ExecuteAsync(CancellationToken stoppingToken)
...
+            var tokenResponse = await client.RequestClientCredentialsTokenAsync(new ClientCredentialsTokenRequest
+            {
+                Address = disco.TokenEndpoint,
+                ClientId = "reportJob",
+                ClientSecret = "secret",
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Jobs/ReportJob.cs:57]: secret is not a secret anymore

File: src/ExadelMentorship.Jobs/ReportJob.cs [hunk 6]
```diff
@@ -42,14 +45,29 @@ protected async override Task ExecuteAsync(CancellationToken stoppingToken)
...
+                Scope = "mailApi"
+            });
+
+            var apiClient = _httpClientFactory.CreateClient();
+            apiClient.SetBearerToken(tokenResponse.AccessToken);
+
             var report = await _reportServices.GenerateReportForUser(reportUser.Id);
             Message message = new Message(reportUser.Email, reportUser.UserName, report);
+            HttpContent httpContent = new StringContent(JsonConvert.SerializeObject(message), Encoding.UTF8, Application.Json);
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.Jobs/ReportJob.cs:66]: Is there a way, i.e. some built-in extension method somewhere that sends C# object in HTTP Post/Put request directly?

I mean it is a common use-case. There must be some convenience way already built-in.

File: src/ExadelMentorship.WebApi/Controllers/MessageController.cs [hunk 7]
```diff
@@ -25,11 +30,18 @@ public async Task SendMail()
             await _emailSender.SendEmailAsync(message);
         }
 
-
+        [Authorize(Policy = "ApiScope")]
```
Comment (kdavletov-exadel) [inline src/ExadelMentorship.WebApi/Controllers/MessageController.cs:33]: Let's make it Role based authorization i.e. It should be authorized if the token contains "sendMail" claim.

**Review by kdavletov-exadel — CHANGES_REQUESTED** [review]

File: src/ExadelMentorship.Auth/IdentityConfiguration.cs (hunk 1 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Auth/IdentityConfiguration.cs:22]: I tested many times, it's not working without. Even in your provided example it exist. 
image

File: src/ExadelMentorship.Auth/IdentityConfiguration.cs (hunk 2 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Auth/IdentityConfiguration.cs:45]: done

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 4 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Jobs/ReportJob.cs:52]: changed

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 5 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Jobs/ReportJob.cs:57]: done

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 3 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Jobs/ReportJob.cs:48]: done

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 6 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Jobs/ReportJob.cs:66]: have not found in google yet

File: src/ExadelMentorship.WebApi/Controllers/MessageController.cs (hunk 7 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.WebApi/Controllers/MessageController.cs:33]: As stackoverflows says clients are not users, so we can't assign roles to a client.
image

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 6 above)
Comment (kdavletov-exadel) [reply src/ExadelMentorship.Jobs/ReportJob.cs:69]: Try `System.Net.Http.Json`

**Review by kdavletov-exadel — CHANGES_REQUESTED** [review]

File: src/ExadelMentorship.Jobs/ReportJob.cs (hunk 6 above)
Comment (SergiLomtadze) [reply src/ExadelMentorship.Jobs/ReportJob.cs]: Thanks, done.

**Review by kdavletov-exadel — APPROVED** [review]

**Pull request merged**